### PR TITLE
Nodal variable bounding using NodalKernels

### PR DESCRIPTION
**Pull request by lindsayad**
Demonstrate imposing both upper and lower bounds on non-linear variables using Lagrange Multipliers, a Nonlinear Complementarity Problem (NCP) function, and `NodalKernels`.

### CODE REVIEW

**Comment by lindsayad** [comment]
I don't understand the app tests...perhaps this is from #14114?

File: framework/include/postprocessors/GreaterThanLessThanPostprocessor.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  PostprocessorValue getValue() override;
+
+private:
+  /// MOOSE variable we compute the contact set from
+  const MooseVariableFEBase & _var;
+
+  /// The libmesh mesh
+  const MeshBase & _mesh;
+
```
Comment (fdkong) [inline framework/include/postprocessors/GreaterThanLessThanPostprocessor.h]: Should store `MooseMesh`, and `MeshBase` can be retrieved from `MooseMesh`

File: test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  [right]
+    type = DirichletBC
+    boundary = right
+    value = 0
+    variable = u
+  []
+[]
+
+[NodalKernels]
+[]
```
Comment (fdkong) [inline test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i]: Do you need this empty `NodalKernels` block? You already have one `NodalKernels` earlier.

File: test/tests/nodalkernels/constraint_enforcement/lower-bound.i [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  [right]
+    type = DirichletBC
+    boundary = right
+    value = 0
+    variable = u
+  []
+[]
+
+[NodalKernels]
+[]
```
Comment (fdkong) [inline test/tests/nodalkernels/constraint_enforcement/lower-bound.i]: Again.

File: test/tests/nodalkernels/constraint_enforcement/upper-bound.i [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  [right]
+    type = DirichletBC
+    boundary = right
+    value = 0
+    variable = u
+  []
+[]
+
+[NodalKernels]
+[]
```
Comment (fdkong) [inline test/tests/nodalkernels/constraint_enforcement/upper-bound.i]: ?

File: test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  []
+[]
+
+[Executioner]
+  type = Transient
+  num_steps = ${num_steps}
+  solve_type = NEWTON
+  dtmin = 1
+  petsc_options_iname = '-snes_max_linear_solve_fail -ksp_max_it -pc_type -sub_pc_factor_levels -snes_linesearch_type'
+  petsc_options_value = '0                           30          asm      16                    basic'
```
Comment (fdkong) [inline test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i]: The default is not zero for `snes_max_linear_solve_fail`?

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
@lindsayad It is a really cool feature! Could you add one more example that has both upper and lower bounds?

File: framework/include/postprocessors/GreaterThanLessThanPostprocessor.h (hunk 1 above)
Comment (lindsayad) [reply framework/include/postprocessors/GreaterThanLessThanPostprocessor.h]: But the object I really want to call methods on is the `MeshBase`. If I don't call `MooseMesh::getMesh` once in the constructor, then I'll be calling it every time the `GreaterThanLessThanPostprocessor::execute` method gets called.

File: test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i (hunk 5 above)
Comment (lindsayad) [reply test/tests/nodalkernels/constraint_enforcement/interpolated-ncp-lm-nodal-enforcement-nodal-forces.i]: No it's not. MOOSE just lets bad linear solves go on forever; it relies on the non-linear solve to not converge

**Review by fdkong — APPROVED** [review]

**Pull request merged**